### PR TITLE
[Protocols-01] implemented protocol Cards

### PR DESCRIPTION
**Pull request by GilCND**

### CODE REVIEW

File: Protocols/CardsDataSource.swift [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+//  CardsDataSource.swift
+//  Protocols
+//
+//  Created by Felipe Gil on 2021-09-24.
+//
+
+import Foundation
+
+protocol CardsDataSource {
+    var playerName: String? { get set }
```
Comment (gpdefreitas) [inline Protocols/CardsDataSource.swift]: We should use `set` for vars that need to be set externally. It looks like that's not the case as your have the `setUserName(player:)` method.

File: Protocols/CardsDataSource.swift [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    var cards: [String] { get set }
+    
+    func generateDeckofCards()
+    func getUserName() -> String
```
Comment (gpdefreitas) [inline Protocols/CardsDataSource.swift]: I know this is common in other languages but in Swift we avoid using `get` in methods as they will necessarily return something.

File: Protocols/DeckofCards.swift [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+//  Protocols
+//
+//  Created by Felipe Gil on 2021-09-24.
+//
+
+import Foundation
+
+
+
+class DeckofCards: CardsDataSource {
```
Comment (gpdefreitas) [inline Protocols/DeckofCards.swift]: Your implementation of the CardsDataSource is fine, but there is room for improvement. Firstly I'd like to ask you what's the reason for making a class here. Also, could you create another data source, now for Spanish cards? And the finally, the most important: do you play Truco?

File: Protocols/CardsDataSource.swift [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+//  Protocols
+//
+//  Created by Felipe Gil on 2021-09-24.
+//
+
+import Foundation
+
+protocol CardsDataSource {
+    var playerName: String? { get set }
+    var cards: [String] { get set }
```
Comment (gpdefreitas) [inline Protocols/CardsDataSource.swift]: Cards also seem to be set internally only as well.

File: Protocols/ViewController.swift [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+//  ViewController.swift
+//  Protocols
+//
+//  Created by Felipe Gil on 2021-09-24.
+//
+
+import UIKit
+
+class ViewController: UIViewController {
+    var deckOfCards = DeckofCards()
```
Comment (gpdefreitas) [inline Protocols/ViewController.swift]: Considering this is user only in the scope of `viewDidLoad` there is not reason to keep it hanging around after that method is done.

**Review by gpdefreitas — COMMENTED** [review]
It looks good for a first protocol-focused exercise, but we can make several improvements. Please take a look at my other comments and let me know what you think.

File: Protocols/CardsDataSource.swift (hunk 1 above)
Comment (GilCND) [reply Protocols/CardsDataSource.swift]: That's right! Thank you!

File: Protocols/CardsDataSource.swift (hunk 2 above)
Comment (GilCND) [reply Protocols/CardsDataSource.swift]: Oh, I see. 
Ok, fixed!

File: Protocols/DeckofCards.swift (hunk 3 above)
Comment (GilCND) [reply Protocols/DeckofCards.swift]: Thinking about it, there is no reason at all. The only "inheritance-ish" that I'm using is from protocols, in fact, more like conformance than inheritance. and I need it by value not reference. 
So I'm updating this to be a Struct. 
I've created some logic inside the same dataSource and created a new one as well. Please let me know which is more suitable.
And for the most relevant question: 6!!!!!!!!!!!!!!!!!!!!

File: Protocols/CardsDataSource.swift (hunk 4 above)
Comment (GilCND) [reply Protocols/CardsDataSource.swift]: Fixed

File: Protocols/ViewController.swift (hunk 5 above)
Comment (GilCND) [reply Protocols/ViewController.swift]: Not sure how to solve this.

**Review by gpdefreitas — APPROVED** [review]

**Pull request merged**